### PR TITLE
perf: precaching & concurrent query

### PR DESCRIPTION
**Pull request by iproudhon**
Added iavl node prefetching & preloading, and made queries concurrent

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the Github PR explorer
- [ ] Review `Codecov Report` in the comment section below once CI passes

### CODE REVIEW

File: store/rootmulti/store.go [hunk 1]
```diff
@@ -404,15 +431,32 @@ func (rs *Store) pruneStores() {
...
 				if errCause := errors.Cause(err); errCause != nil && errCause != iavltree.ErrVersionDoesNotExist {
 					panic(err)
 				}
 			}
 		}
 	}
 
-	rs.pruneHeights = make([]int64, 0)
+	// remove pruned heights from rs.pruneHeights
+	// new rs.pruneHeights will be updated by next flushMedataData cycle
```
Comment (dudong2) [inline store/rootmulti/store.go:446]: typo?: flushMedataData -> flushMetaData

File: store/iavl/cache.go [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+	return &ristrettoCache{
+		Cache: cache,
+	}
+}
+
+func (c *ristrettoCache) Set(key, value []byte) {
+	c.Cache.Set(key, value, int64(len(value)))
+}
+
+func (c *ristrettoCache) Get(dst, key []byte) []byte {
```
[the remark this answers is not in the document]
Comment (iproudhon) [reply store/iavl/cache.go:48]: It's to comply with existing fastcache api.

File: store/iavl/fixedmap.go [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+	gets    int64
+	hits    int64
+	misses  int64
+}
+
+func NewFixedMap(maxEntries int) *FixedMap {
+	return &FixedMap{
+		maxEntries: uint64(maxEntries),
+		maps:       make([][]byte, maxEntries),
+		locks:      make([]sync.Mutex, 1009),
```
[the remark this answers is not in the document]
Comment (iproudhon) [reply store/iavl/fixedmap.go:28]: It's random prime number.

File: store/iavl/cache.go [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+	return &freeCache{
+		Cache: freecache.NewCache(cacheSize),
+	}
+}
+
+func (c *freeCache) Set(key, value []byte) {
+	c.Cache.Set(key, value, 0)
+}
+
+func (c *freeCache) Get(dst, key []byte) []byte {
```
[the remark this answers is not in the document]
Comment (iproudhon) [reply store/iavl/cache.go:88]: It's to comply with existing fastcache api.

File: store/rootmulti/store.go (hunk 1 above)
Comment (iproudhon) [reply store/rootmulti/store.go]: Fixed

**Comment by iproudhon** [comment]
Changes are already merged in.

**Pull request closed without merging**